### PR TITLE
nGraph Python API tweaks and tutorial

### PR DESCRIPTION
**Pull request by postrational**

### CODE REVIEW

File: ngraph/python/TUTORIAL.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        print('Operation {} of type {} has no attributes.'.format(node.get_friendly_name(), node.get_type_name()))
+```
+
+##### Node runtime information
+
+Attributes are properties of nodes in the computational graph, which will be stored when the model is serialized to a file. However, there can be additional properties of nodes, which are only important during the execution of a graph. An example of such a property is `affinity`, which determines which operation will be executed on which hardware in a heterogeneous environment. You can get and set runtime information by using the `get_rt_info` method of a node.
+
+```python
+rt_info = node.get_rt_info()
+    rt_info["affinity"] = "test_affinity"
```
Comment (jiwaszki) [inline ngraph/python/TUTORIAL.md:192]: Small formatting error here, besides that legitimate.

**Review by jiwaszki — CHANGES_REQUESTED** [review]

**Review by mbencer — APPROVED** [review]

File: ngraph/python/TUTORIAL.md (hunk 1 above)
Comment (postrational) [reply ngraph/python/TUTORIAL.md]: Fixed

File: ngraph/python/TUTORIAL.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    y = ReLU(x + b)
+
+You can use the following code to create parameters `x` and `b` and create the computational graph representing the formula:
+
+```python
+import numpy as np
+import ngraph as ng
+
+parameter_x = ng.parameter(shape=[2, 2], dtype=np.float32, name="x")
```
Comment (GlebKazantaev) [inline ngraph/python/TUTORIAL.md:28]: In C++ documentation we use tags like this https://github.com/openvinotoolkit/openvino/blob/master/docs/examples/example_ngraph_utils.cpp#L9 to copy code snippets from source code to documentation. In C++ it helps to avoid syntax errors because we build this code. But in case of python I'm not sure that this tags will help.

File: ngraph/python/TUTORIAL.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+ <Relu: 'y' ({2,2})>,
+ <Result: 'Result_4' ({2,2})>]
+```
+
+Each `Node` has a unique `name` property, assigned at creation time. User provided names can be retrieved using the `get_friendly_name` method. `get_type_name` returns the operation type of the node and the `shape` property returns the shape of the node's output tensor.
+
+```python
+>>> for node in function.get_ordered_ops():
+>>>     print('Node name: {:15} Friendly name: {:10} Op: {:10} {}'.format(
+>>>         node.name, node.get_friendly_name(), node.get_type_name(), node.shape))
```
Comment (GlebKazantaev) [inline ngraph/python/TUTORIAL.md:118]: Do we need to say smth about dynamic shapes here?

File: ngraph/python/TUTORIAL.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    print('{} node has a static shape {}'.format(node.get_friendly_name(), shape))
+```
+
+##### Node attributes
+
+Nodes may also have additional attributes. For example a `Softmax` node has an `axis` attribute. Each attribute may be retrieved using a dedicated method, for example `node.get_axis()`.
+
+```python
+>>> node = ng.softmax(data=[[1, 2], [3, 4]], axis=1)
+>>> node.get_axis()
```
Comment (GlebKazantaev) [inline ngraph/python/TUTORIAL.md:167]: I think we need to add some words about set_XXX methods.

File: ngraph/python/TUTORIAL.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        print('Operation {} of type {} has attributes:'.format(node.get_friendly_name(), node.get_type_name()))
+        for attr, value in attributes.items():
+            print("  * {}: {}".format(attr, value))
+    else:
+        print('Operation {} of type {} has no attributes.'.format(node.get_friendly_name(), node.get_type_name()))
+```
+
+##### Node runtime information
+
+Attributes are properties of nodes in the computational graph, which will be stored when the model is serialized to a file. However, there can be additional properties of nodes, which are only important during the execution of a graph. An example of such a property is `affinity`, which determines which operation will be executed on which hardware in a heterogeneous environment. You can get and set runtime information by using the `get_rt_info` method of a node.
```
Comment (GlebKazantaev) [inline ngraph/python/TUTORIAL.md:188]: If we want to keep this section we need to say that rt info attribute are currently used only for internal needs and we do not guaranty backward compatibility for this attributes.

File: ngraph/python/TUTORIAL.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+You can retrieve the data tensor associated with a `Constant` node by calling its `get_data` method:
+
+```python
+if node.get_type_name() == 'Constant':
+    data = node.get_data()
+``` 
+
+#### Optimizing graphs
+
+nGraph supports a growing number of optimization passes, which can be applied to a graph to increase model execution performance. In order to perform an optimization such as `ConstantFolding`, you will need to create an optimization pass `Manager`, register the passes you want to execute and feed an nGraph function to the configured pass manager.
```
Comment (GlebKazantaev) [inline ngraph/python/TUTORIAL.md:206]: ConstantFolding is not an optimization pass and it potentially can break model reshapebility. So I would suggest to replace word "optimization" with "constant sub-graphs folding".

File: ngraph/python/TUTORIAL.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+The following code example illustrates the principal:
+
+```python
+from ngraph.impl.passes import Manager
+pass_manager = Manager()
+pass_manager.register_pass('ConstantFolding')
+pass_manager.set_per_pass_validation(False)
```
Comment (GlebKazantaev) [inline ngraph/python/TUTORIAL.md:214]: It was used in MA to workaround some issues, so I would suggest to remove this line from example.

File: ngraph/python/TUTORIAL.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+import numpy as np
+import ngraph as ng
+
+parameter_x = ng.parameter(shape=[2, 2], dtype=np.float32, name="x")
+parameter_b = ng.parameter(shape=[2, 2], dtype=np.float32, name="b")
+
+graph = ng.relu(parameter_x + parameter_b, name="y")
+```
+
+nGraph opertions are organized into sets, which are released with new versions of OpenVINO. As operation signatures can change between operation sets, it's good practice to use operations from a specific operation set in your scripts.
```
Comment (GlebKazantaev) [inline ngraph/python/TUTORIAL.md:34]: "opertions" -> "operations"

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Comment by GlebKazantaev** [comment]
@postrational you need to create separate PR to release branch.

@andrew-zaytsev could you tell us how this md file must be connected to other OV documentation?

**Review by jiwaszki — APPROVED** [review]

File: ngraph/python/TUTORIAL.md (hunk 3 above)
Comment (postrational) [reply ngraph/python/TUTORIAL.md:118]: Dynamic shapes are discussed further down.

File: ngraph/python/TUTORIAL.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (postrational) [reply ngraph/python/TUTORIAL.md:34]: I added language to make this clear.

File: ngraph/python/TUTORIAL.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+```python
+import ngraph as ng
+dir(ng)
+help(ng.space_to_batch)
+```
+
+
+#### Creating a simple graph
+
+nGraph's Python API can be used to create computational models. Let's say you want to create a model which represents the following formula:
```
[the remark this answers is not in the document]
Comment (postrational) [reply ngraph/python/TUTORIAL.md:18]: I added language to make this clear, I don't think we need to switch order of the examples.

File: ngraph/python/TUTORIAL.md (hunk 4 above)
Comment (postrational) [reply ngraph/python/TUTORIAL.md:167]: Done.

File: ngraph/python/TUTORIAL.md (hunk 5 above)
Comment (postrational) [reply ngraph/python/TUTORIAL.md:188]: Done.

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
+# Using nGraph's Python API
+
+nGraph is OpenVINO's graph manipulation library, used to represent neural network models in the form of a computational graph. Using nGraph's Python APIs, you can create, inspect and manipulate computational graphs. You can also use the API to set runtime properties of the graph, which will be used by OpenVINO during inference.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:3]: nGraph is the OpenVINO™ graph manipulation library that is used to represent neural network models in the form of a computational graph. With nGraph Python APIs, you can create, inspect, and manipulate computational graphs. ...

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+## nGraph namespace module
+
+nGraph has a large set of mathematical operations, chosen to cover most ops used in popular neural network models. The `ngraph` module is the basic namespace used to expose commonly used API methods. You can inspect the `ngraph` module or call `help` on any method for usage information.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:7]: most operations

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+```python
+import ngraph as ng
+dir(ng)
+help(ng.space_to_batch)
+```
+
+## Create a simple graph
+
+nGraph's Python API can be used to create computational models. Let's say you want to create a model which represents the following formula:
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:17]: You can use nGraph Python API to create computational models. For example, to create a model that represents the following formula y=ReLU(x + b), you can use the following code to create parameters `x` and `b` and create the computational graph:

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+import numpy as np
+import ngraph as ng
+
+parameter_x = ng.parameter(shape=[2, 2], dtype=np.float32, name="x")
+parameter_b = ng.parameter(shape=[2, 2], dtype=np.float32, name="b")
+
+graph = ng.relu(parameter_x + parameter_b, name="y")
+```
+
+nGraph operations are organized into sets, which are released with new versions of OpenVINO. As operation signatures can change between operation sets, it's good practice to use operations from a specific operation set in your scripts. This will ensure that your code is compatible with future versions of OpenVINO.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:33]: it is a good practice

Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:33]: This ensures that your code...

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+parameter_x = ops.parameter(shape=[2, 2], dtype=np.float32, name="x")
+parameter_b = ops.parameter(shape=[2, 2], dtype=np.float32, name="b")
+
+graph = ops.relu(parameter_x + parameter_b, name="y")
+```
+
+
+## Create an nGraph function from a graph
+
+You can combine a graph together with it's input parameters into an nGraph function object. An nGraph function can be passed to other parts of OpenVINO to perform analysis and inference.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:49]: its input parameters

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+
+```python
+>>> function = ng.Function(graph, [parameter_x, parameter_b], "TestFunction")
+<Function: 'TestFunction' ({2,2})>
+```
+
+## Run inference
+
+In order to run inference on an nGraph function, you'll need to convert it to an Inference Engine network
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:58]: To run inference on an nGraph function, convert it to an Inference Engine network

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+>>>    'b': np.array([[-1, 0], [0, 1]]),
+>>> })
+>>> request.output_blobs['y'].buffer
+array([[-0.,  0.],
+       [ 1.,  3.]], dtype=float32)
+```
+
+## Load an nGraph function from a file
+
+It's common to load a model from a file using the Inference Engine `read_network` method:
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:80]: You can load a model from a file using the Inference Engine...

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+network = ie.read_network(model=model_path)
+
+import ngraph as ng
+function = ng.function_from_cnn(network)
+```
+
+
+## Inspect an nGraph function
+
+You can use the nGraph function's `get_ordered_ops` method to get a topologically sorted list of it's graph `Node`s.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:101]: its graph

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+```python
+>>> function.get_ordered_ops()
+[<Parameter: 'b' ({2,2}, float)>,
+ <Parameter: 'x' ({2,2}, float)>,
+ <Add: 'Add_2' ({2,2})>,
+ <Relu: 'y' ({2,2})>,
+ <Result: 'Result_4' ({2,2})>]
+```
+
+Each `Node` has a unique `name` property, assigned at creation time. User provided names can be retrieved using the `get_friendly_name` method. `get_type_name` returns the operation type of the node and the `shape` property returns the shape of the node's output tensor.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:112]: User-provided names

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+
+```python
+>>> for node_input in node.inputs():
+>>>     parent = node_input.get_source_output().get_node()
+>>>     print('{} node has parent {}'.format(node.get_friendly_name(), parent.get_friendly_name()))
+Add_2 node has parent x
+Add_2 node has parent b
+```
+
+By analogy, `Node`'s children can be retrieved using the `outputs()` method which returns a list of `Output` objects. By contrast to inputs, each output can be connected to multiple downstream nodes.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:137]: method, which returns

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+>>> for node_output in node.outputs():
+>>>     children = [target.get_node() for target in node_output.get_target_inputs()]
+>>>     print('{} node has children {}'.format(node.get_friendly_name(), ','.join([c.get_friendly_name() for c in children])))
+>>> 
+Add_2 node has children y
+```
+
+### Shapes
+
+Each `Input` and `Output` has an associated tensor shape, which describes what data it can accept or produce. Shapes can be dynamic, which means that they are only partially known. You can get information about this shape by calling the `get_partial_shape` method. If the returned `PartialShape` object's `is_dynamic` property returns `True`, then the shape is known partially. Otherwise you can convert the shape to a fully known static shape (list of integers).
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:149]: Otherwise,

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+if partial_shape.is_dynamic:
+    print('{} node has a dynamic shape {}'.format(node.get_friendly_name(), partial_shape))
+else:
+    shape = list(partial_shape.to_shape())
+    print('{} node has a static shape {}'.format(node.get_friendly_name(), shape))
+```
+
+### Node attributes
+
+Nodes may also have additional attributes. For example a `Softmax` node has an `axis` attribute. Each attribute may be retrieved using a dedicated method, for example `node.get_axis()`.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:162]: For example, a `Softmax` node...
for example,

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+You can retrieve the data tensor associated with a `Constant` node by calling its `get_data` method:
+
+```python
+if node.get_type_name() == 'Constant':
+    data = node.get_data()
+``` 
+
+## Optimize graphs
+
+nGraph supports a growing number of optimization passes, which can be applied to a graph to increase model execution performance. In order to perform constant folding of sub-graphs you can use the `ConstantFolding` pass. To do this you will need to create an optimization pass `Manager`, register the passes you want to execute and feed an nGraph function to the configured pass manager.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:213]: To perform constant folding of subgraphs, you can use the `ConstantFolding` pass. To do this, create an optimization pass `Manager`, register the passes you need to execute, and feed an nGraph function to the configured pass manager.

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+
+The following code example illustrates the principal:
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:215]: illustrates the process:

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 24]
```diff
@@ -55,7 +52,7 @@ You can combine a graph together with it's input parameters into an nGraph funct
 
 ## Run inference
 
-In order to run inference on an nGraph function, you'll need to convert it to an Inference Engine network
+In order to run inference on an nGraph function convert it to an Inference Engine network and call its `infer` method.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraph_Python_API.md:55]: please add comma before convert

File: docs/nGraph_DG/nGraph_Python_API.md [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+You can retrieve the data tensor associated with a `Constant` node by calling its `get_data` method:
+
+```python
+if node.get_type_name() == 'Constant':
+    data = node.get_data()
+``` 
+
+## Optimize graphs
+
+nGraph supports a growing number of optimization passes, which can be applied to a graph to increase model execution performance. To perform constant folding of subgraphs, you can use the `ConstantFolding` pass. To do this, create an optimization pass `Manager`, register the passes you need to execute, and feed an nGraph function to the configured pass manager.
```
Comment (GlebKazantaev) [inline docs/nGraph_DG/nGraph_Python_API.md:209]: We do not suggest user to call any transformations to increase model performance neither in C++ nor in Python API. All available transformations required usually for some advanced cases.
So I suggest to rename this section to `Apply transformations` and replace `nGraph supports a growing number of optimization passes, which can be applied to a graph to increase model execution performance.` with smth like `nGraph provides some transformations that could be applied to ngraph::Function and usually used for  internal needs.` And also I would add `Please note that the list of available transformations may change in future versions of OpenVINO.`

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Review by GlebKazantaev — APPROVED** [review]

**Review by avladimi — APPROVED** [review]

**Review by andrew-zaytsev — APPROVED** [review]

**Review by andrew-zaytsev — APPROVED** [review]
